### PR TITLE
First implementation of single-GPU FFT using cuFFT

### PR DESCRIPTION
**Pull request by mferreravila**
Add support for FFTs in single-GPU using cuFFT as back-end.

### CODE REVIEW

File: cunumeric/fft/fft.py
Comment (magnatelee) [inline cunumeric/fft/fft.py:88]: Change this line to `Single GPU`. (Same for all the other docstrings.)

File: src/cunumeric.mk
Comment (magnatelee) [inline src/cunumeric.mk:129]: Make sure this line is aligned with others.

File: cunumeric/array.py
Comment (magnatelee) [inline cunumeric/array.py:2240]: What's the reason for the multiplication with 1.0? You may think that looks innocuous, but that will launch tasks. If what you wanted is to multiply `out` with `1.0 / factor`, you may want to put parentheses to make sure of it.

File: cunumeric/config.py
Comment (magnatelee) [inline cunumeric/config.py:252]: This entire class looks like a C code. You could make this object oriented so that the if conditions you're checking in the methods and properties are reflected in the class hierarchy. For example, you could make a class for each FFT type (say R2C, C2R, C2C, D2Z, Z2D, and Z2Z) and have them implement the same set of methods and properties. Then each will need to return the right value without checking anything. For instance, `R2C.input_dtype` would just return `np.float32` unconditionally. Of course, you will still need constructor methods like `real_to_complex_code` or `complex_to_real_code`, but you can eliminate the need for checking the FFT codes once you created them.

File: cunumeric/deferred.py
Comment (magnatelee) [inline cunumeric/deferred.py:811]: can `axes` ever be `None`? as far as I can tell, that case is already pruned out by the frontend...

File: src/cunumeric/fft/fft.h
Comment (magnatelee) [inline src/cunumeric/fft/fft.h:38]: Please remove any variant that you didn't implement. Having an empty body would only make the mapper confused. The contract is that all variants must do the same computation.

File: src/cunumeric/fft/fft.h
Comment (magnatelee) [inline src/cunumeric/fft/fft.h:40]: Same here.

File: src/cunumeric/fft/fft.h
Comment (magnatelee) [inline src/cunumeric/fft/fft.h:1]: I believe this year is 2022.

File: src/cunumeric/fft/fft_util.h
Comment (magnatelee) [inline src/cunumeric/fft/fft_util.h:27]: So this enum is decoupled from the enum in the Python code, but I think you should use the same enum to avoid any mistakes. Please look at how `CuNumericOpCode` is written.

File: tests/fft_c2c.py
Comment (magnatelee) [inline tests/fft_c2c.py:36]: I guess you could use a for loop to iterate the test vectors, instead of enumerating them like this.

File: tests/fft_c2c.py
Comment (magnatelee) [inline tests/fft_c2c.py:78]: Same here. You could have saved a lot of boilerplates.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:62]: Any reason that you didn't use the plan cache? If the cache doesn't give you what you want, you could either extend it your self or ask me to do it.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:73]: Let's use `create_buffer`: https://github.com/nv-legate/legate.core/blob/branch-22.05/src/core/data/buffer.h#L26-L29

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:91]: Please remove these comments before your PR gets merged.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:93]: Legion's `DeferredBuffer` is re-exported by Legate on the name of `legate::Buffer`. So you can just type `Buffer<OUTPUT_TYPE, DIM>`.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:101]: I don't think you need these castings to `void*`. That would happen implicitly. If you insist on casting them, please use `static_cast`. (I saw there are other places like this, so I'd like you to fix them all.)

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:171]: Again, let's use `create_buffer`.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:182]: Again, let me know how the existing plan cache can replace this code.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:236]: You don't need this synchronization, as the runtime will do it for you.

File: src/cunumeric/fft/fft.cu
Comment (magnatelee) [inline src/cunumeric/fft/fft.cu:241]: This function looks quite similar to `cufft_over_axis_c2c`. Can you think of a way to factor out the common parts?

File: cunumeric/array.py
Comment (mferreravila) [reply cunumeric/array.py:2132]: Not really. axes = [-5, 10] is not a valid input and it should error out before sanitizing. Unless there is a compelling reason to deviate from numpy behavior?

File: cunumeric/array.py
Comment (mferreravila) [reply cunumeric/array.py:2196]: From your comments on slack
(Feb 16):
```
what I’d do for a functional implementation is to make a padded array, fill it with zeros, slice it to a sub-array whose shape matches the original array’s, and do a copy between the two
```
(Feb 24):
```
for zero padding, I think you can do something like this:
1. create an empty deferred array A of the size
2. call fill to zero it out
3. slice A to match the shape of s and call copy
```

Tests fail without the copy. But if this copy is redundant, I'd be happy to amend it with your feedback. Maybe the thunk is not being assigned correctly?

Comment (magnatelee) [reply cunumeric/array.py:2196]: I think I got confused by the fact that lines 2180-2183 already make a copy of the input. I think the copy at line 2193 should be moved to that if statement to make sure we make only one copy of the input.

File: src/cunumeric/fft/fft.h
Comment (mferreravila) [reply src/cunumeric/fft/fft.h:1]: :slightly_smiling_face:

File: cunumeric/config.py
Comment (mferreravila) [reply cunumeric/config.py:263]: What constructor arguments would you suggest?

I'll probably leave this to you, as you seem to have a clear idea of how you'd like these classes to look, and any changes on my end might steer this again into C++ territory.

Comment (magnatelee) [reply cunumeric/config.py:263]: Ok. I'll approve this PR and make some follow-up changes if you don't mind me doing it.

**Review by magnatelee — APPROVED** [review]

**Comment by magnatelee** [comment]
@mferreravila Like I said in the other comment, I'll take this over from you and polish it up, unless you're interested in finishing it up yourself. Just let me know.

**Pull request merged**
